File: phases/12-cryptography-and-security/10-public-key-ii-rsa-internals-padding/code/main.py
"""
Public Key II — RSA Internals & Padding
Phase 12 — Cryptography & Security

Working RSA implementation with key generation (Miller-Rabin),
textbook encryption/decryption, and OAEP padding.
"""

import hashlib
import os
import random


def is_prime(n: int, k: int = 10) -> bool:
    if n < 2:
        return False
    if n < 4:
        return True
    if n % 2 == 0:
        return False
    r, d = 0, n - 1
    while d % 2 == 0:
        r += 1
        d //= 2
    for _ in range(k):
        a = random.randrange(2, n - 2)
        x = pow(a, d, n)
        if x == 1 or x == n - 1:
            continue
        for _ in range(r - 1):
            x = pow(x, 2, n)
            if x == n - 1:
                break
        else:
            return False
    return True


def generate_prime(bits: int) -> int:
    while True:
        n = int.from_bytes(os.urandom(bits // 8), "big")
        n |= (1 << (bits - 1)) | 1
        if is_prime(n, k=20):
            return n


def generate_keypair(bits: int = 512):
    p = generate_prime(bits // 2)
    q = generate_prime(bits // 2)
    n = p * q
    phi = (p - 1) * (q - 1)
    e = 65537
    d = pow(e, -1, phi)
    return (n, e, p, q), (n, d)


def encrypt(pub_key: tuple, plaintext: int) -> int:
    n, e = pub_key[0], pub_key[1]
    return pow(plaintext, e, n)


def decrypt(priv_key: tuple, ciphertext: int) -> int:
    n, d = priv_key
    return pow(ciphertext, d, n)


def mgf1(seed: bytes, length: int, hash_func=hashlib.sha1) -> bytes:
    hlen = hash_func().digest_size
    result = b""
    counter = 0
    while len(result) < length:
        c = counter.to_bytes(4, "big")
        result += hash_func(seed + c).digest()
        counter += 1
    return result[:length]


def oaep_pad(message: bytes, n_bytes: int) -> bytes:
    import hashlib
    hash_func = hashlib.sha1
    hlen = hash_func().digest_size
    max_msg_len = n_bytes - 2 * hlen - 2
    if len(message) > max_msg_len:
        raise ValueError(f"Message too long (max {max_msg_len} bytes)")

    label_hash = hash_func(b"").digest()
    ps_len = n_bytes - len(message) - 2 * hlen - 2
    ps = b"\x00" * ps_len
    db = label_hash + ps + b"\x01" + message

    seed = os.urandom(hlen)
    db_mask = mgf1(seed, n_bytes - hlen - 1)
    masked_db = bytes(a ^ b for a, b in zip(db, db_mask))
    seed_mask = mgf1(masked_db, hlen)
    masked_seed = bytes(a ^ b for a, b in zip(seed, seed_mask))
    return b"\x00" + masked_seed + masked_db


def oaep_unpad(padded: bytes, n_bytes: int) -> bytes:
    import hashlib
    hash_func = hashlib.sha1
    hlen = hash_func().digest_size

    if len(padded) != n_bytes:
        raise ValueError("Decryption error")
    if padded[0] != 0:
        raise ValueError("Decryption error")

    masked_seed = padded[1:1 + hlen]
    masked_db = padded[1 + hlen:]

    seed_mask = mgf1(masked_db, hlen)
    seed = bytes(a ^ b for a, b in zip(masked_seed, seed_mask))
    db_mask = mgf1(seed, n_bytes - hlen - 1)
    db = bytes(a ^ b for a, b in zip(masked_db, db_mask))

    label_hash = hash_func(b"").digest()
    if db[:hlen] != label_hash:
        raise ValueError("Decryption error")

    i = hlen
    while i < len(db) and db[i] == 0:
        i += 1
    if i >= len(db) or db[i] != 1:
        raise ValueError("Decryption error")

    return db[i + 1:]


def encrypt_oaep(pub_key: tuple, message: bytes) -> int:
    n = pub_key[0]
    n_bytes = (n.bit_length() + 7) // 8
    padded = oaep_pad(message, n_bytes)
    m = int.from_bytes(padded, "big")
    return encrypt(pub_key, m)


def decrypt_oaep(priv_key: tuple, ciphertext: int) -> bytes:
    n = priv_key[0]
    n_bytes = (n.bit_length() + 7) // 8
    padded_int = decrypt(priv_key, ciphertext)
    padded = padded_int.to_bytes(n_bytes, "big")
    return oaep_unpad(padded, n_bytes)


def main() -> None:
    print("=== RSA Key Generation (512-bit demo) ===")
    pub_key, priv_key = generate_keypair(bits=512)
    n, e, p, q = pub_key
    print(f"p (prime 1):        {p}")
    print(f"q (prime 2):        {q}")
    print(f"n (modulus):        {n}")
    print(f"e (public exp):     {e}")
    print(f"d (private exp):    {priv_key[1]}")
    print(f"Bit length:         {n.bit_length()}")

    print("\n=== Textbook RSA Encryption/Decryption ===")
    message_int = 42
    cipher = encrypt(pub_key, message_int)
    plain = decrypt(priv_key, cipher)
    print(f"Message:              {message_int}")
    print(f"Ciphertext:           {cipher}")
    print(f"Decrypted:            {plain}")
    print(f"Match:                {message_int == plain}")

    print("\n=== Textbook RSA is Deterministic ===")
    c1 = encrypt(pub_key, 42)
    c2 = encrypt(pub_key, 42)
    print(f"Encrypt(42) again:    {c2}")
    print(f"Same ciphertext:      {c1 == c2}  (BAD!)")

    print("\n=== Textbook RSA is Malleable ===")
    c = encrypt(pub_key, 100)
    modified = (c * pow(2, e, n)) % n
    decrypted = decrypt(priv_key, modified)
    print(f"Encrypt(100):         {c}")
    print(f"Modified ciphertext:  {modified}")
    print(f"Decrypted modified:   {decrypted} (should be 200)")

    print("\n=== RSA-OAEP Encryption ===")
    message = b"Hi, OAEP!"
    ct = encrypt_oaep(pub_key, message)
    pt = decrypt_oaep(priv_key, ct)
    print(f"Original message:     {message}")
    print(f"Ciphertext (int):     {ct}")
    print(f"Decrypted message:    {pt}")
    print(f"Match:                {message == pt}")

    print("\n=== RSA-OAEP is Non-Deterministic ===")
    ct2 = encrypt_oaep(pub_key, message)
    print(f"Same message:         {ct == ct2}  (should be False — OAEP uses random seed)")

    print("\n=== RSA-OAEP Tamper Detection ===")
    tampered = ct ^ 1
    try:
        decrypt_oaep(priv_key, tampered)
        print("Tampered ciphertext decrypted (BAD!)")
    except ValueError as e:
        print(f"Tampered ciphertext rejected: {e}")


if __name__ == "__main__":
    main()

File: phases/12-cryptography-and-security/10-public-key-ii-rsa-internals-padding/code/main.rs
// Public Key II — RSA Internals & Padding
// Phase 12 — Cryptography & Security
//
// RSA math implementation with 64-bit integers for educational
// demonstration: modular exponentiation, Miller-Rabin primality
// test, extended Euclidean algorithm, key generation, encryption,
// and decryption.
//
// Note: 64-bit primes are trivially factorable. Real RSA uses
// 2048+ bit moduli. This code demonstrates the algorithm's
// core mechanics with small numbers.

use std::time::{SystemTime, UNIX_EPOCH};

/// Modular exponentiation: (base^exp) % modulus
/// Uses square-and-multiply with u128 intermediates to avoid overflow.
fn mod_pow(mut base: u64, mut exp: u64, modulus: u64) -> u64 {
    let mut result = 1u64;
    base %= modulus;
    while exp > 0 {
        if exp & 1 == 1 {
            result = (result as u128 * base as u128 % modulus as u128) as u64;
        }
        base = (base as u128 * base as u128 % modulus as u128) as u64;
        exp >>= 1;
    }
    result
}

/// Miller-Rabin primality test (deterministic for n < 2^64).
/// Uses the known set of bases that are sufficient for 64-bit integers:
/// [2, 3, 5, 7, 11, 13, 17] — no false positives for n < 2^64.
fn is_prime(n: u64) -> bool {
    if n < 2 { return false; }
    if n % 2 == 0 { return n == 2; }

    // Write n-1 as d * 2^r
    let mut d = n - 1;
    let mut r = 0;
    while d % 2 == 0 {
        d /= 2;
        r += 1;
    }

    // Bases sufficient for deterministic test on u64
    let bases: [u64; 7] = [2, 3, 5, 7, 11, 13, 17];

    'next_base: for &a in &bases {
        if a >= n { continue; }
        let mut x = mod_pow(a, d, n);
        if x == 1 || x == n - 1 { continue; }

        for _ in 0..r - 1 {
            x = mod_pow(x, 2, n);
            if x == n - 1 { continue 'next_base; }
        }
        return false;
    }
    true
}

/// Simple LCG for generating pseudo-random values (not cryptographically secure).
fn lcg(seed: &mut u64) -> u64 {
    *seed = seed.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
    *seed
}

/// Generate a random prime of approximately `bits` bits.
fn generate_prime(bits: u32, seed: &mut u64) -> u64 {
    loop {
        let mut n = lcg(seed) >> (64 - bits);
        n |= 1 << (bits - 1);  // Ensure bit length
        n |= 1;                // Ensure odd
        if is_prime(n) {
            return n;
        }
    }
}

/// Extended Euclidean algorithm returning (gcd, x, y) where a*x + b*y = gcd.
fn extended_gcd(a: i64, b: i64) -> (i64, i64, i64) {
    if a == 0 { return (b, 0, 1); }
    let (g, x1, y1) = extended_gcd(b % a, a);
    (g, y1 - (b / a) * x1, x1)
}

/// Compute modular inverse of a modulo m (m must be prime or a and m coprime).
fn mod_inverse(a: u64, m: u64) -> u64 {
    let (g, x, _) = extended_gcd(a as i64, m as i64);
    assert_eq!(g, 1, "a and m are not coprime");
    ((x % m as i64 + m as i64) % m as i64) as u64
}

/// Generate an RSA keypair: returns (n, e, d, p, q).
fn generate_keypair(bits: u32, seed: &mut u64) -> (u64, u64, u64, u64, u64) {
    let p = generate_prime(bits, seed);
    let q = generate_prime(bits, seed);
    let n = p * q;
    let phi = (p - 1) * (q - 1);
    let e: u64 = 65537;
    let d = mod_inverse(e, phi);
    (n, e, d, p, q)
}

/// RSA encryption: c = m^e mod n
fn encrypt(m: u64, e: u64, n: u64) -> u64 {
    mod_pow(m, e, n)
}

/// RSA decryption: m = c^d mod n
fn decrypt(c: u64, d: u64, n: u64) -> u64 {
    mod_pow(c, d, n)
}

/// Demonstrate textbook RSA with small parameters.
fn demonstrate_rsa() {
    println!("=== Textbook RSA (16-bit primes) ===");

    let mut seed = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_nanos() as u64;

    let (n, e, d, p, q) = generate_keypair(10, &mut seed);

    println!("p = {}", p);
    println!("q = {}", q);
    println!("n = {} ({} bits)", n, (n as f64).log2().ceil() as u32);
    println!("e = {}", e);
    println!("d = {}", d);

    // Encrypt and decrypt a small message
    let message = 42u64;
    let cipher = encrypt(message, e, n);
    let plain = decrypt(cipher, d, n);

    println!("\nMessage:    {}", message);
    println!("Ciphertext: {}", cipher);
    println!("Decrypted:  {}", plain);
    println!("Match:      {}\n", message == plain);

    // Demonstrate that RSA is deterministic
    let c1 = encrypt(42, e, n);
    let c2 = encrypt(42, e, n);
    println!("Deterministic: encrypt(42) twice = {} and {}", c1, c2);
    println!("  Same? {} (BAD — should be different with padding)\n", c1 == c2);

    // Demonstrate malleability
    let c = encrypt(100, e, n);
    let mult = mod_pow(2, e, n);
    let modified = (c as u128 * mult as u128 % n as u128) as u64;
    let decrypted = decrypt(modified, d, n);
    println!("Malleability: encrypt(100) = {}", c);
    println!("  Modified ciphertext decrypts to: {} (should be 200)\n", decrypted);
}

/// Demonstrate RSA with a larger (but still tiny) key.
fn demonstrate_larger_key() {
    println!("=== RSA with 20-bit primes ===");

    let mut seed = 123456789u64;
    let (n, e, d, p, q) = generate_keypair(12, &mut seed);

    println!("p = {}", p);
    println!("q = {}", q);
    println!("n = {} ({} bits)", n, (n as f64).log2().ceil() as u32);

    let message = 12345u64;
    let cipher = encrypt(message, e, n);
    let plain = decrypt(cipher, d, n);
    println!("Message: {}, Decrypted: {}, Match: {}", message, plain, message == plain);
    println!();
}

/// Benchmark modular exponentiation with different exponents.
fn benchmark_mod_pow() {
    println!("=== Exponentiation Benchmark ===");
    let modulus = 65521u64; // A real prime
    for exp_bits in [8, 12, 16, 20] {
        let exp = (1u64 << exp_bits) - 1;
        let start = std::time::Instant::now();
        let iterations = 1000;
        for _ in 0..iterations {
            let _ = mod_pow(2, exp, modulus);
        }
        let elapsed = start.elapsed();
        let avg = elapsed / iterations;
        println!("mod_pow(2, 2^{}-1, 65521): avg {:?}", exp_bits, avg);
    }
}

fn main() {
    demonstrate_rsa();
    demonstrate_larger_key();
    benchmark_mod_pow();

    // Demonstrate Euler's theorem: a^phi(n) ≡ 1 (mod n) for coprime a, n
    println!("=== Euler's Theorem Verification ===");
    let p = 101u64;
    let q = 103u64;
    let n = p * q;
    let phi = (p - 1) * (q - 1);
    let a = 5u64;  // coprime to n
    let result = mod_pow(a, phi, n);
    println!("5^{} mod {} = {} (should be 1)\n", phi, n, result);

    // Show that e*d ≡ 1 mod phi
    println!("=== RSA Key Relationship ===");
    let e = 65537u64;
    let d = mod_inverse(e, phi);
    let check = (e as u128 * d as u128) % phi as u128;
    println!("e * d mod phi = {} (should be 1)", check);
    println!("Public:  (n={}, e={})", n, e);
    println!("Private: (n={}, d={})", n, d);

    // End-to-end with this key
    let msg = 2024u64;
    let ct = encrypt(msg, e, n);
    let pt = decrypt(ct, d, n);
    println!("Message: {}, Cipher: {}, Decrypted: {}, OK: {}", msg, ct, pt, msg == pt);
}
